### DISCUSSION TITLE
More form handling power in Axum?

### DISCUSSION THREAD

**Discussion by sidkshatriya** [tokio-rs/axum#295] [Q&A] [answered]
Are there some third party libraries I can use for form handling in Axum? The current form handling example seems to be quite basic and I'm looking for a lot more form handling power. For instance, Rocket seems to have powerful form handling capabilities.

**Comment by davidpdrsn** [comment, marked as answer]
Can you elaborate on what features you're looking for?

In general we try not to build everything into axum directly but rather make it easy to provide integrations that work with other crates. That way you get to pick/build the implementation that is right for you and we avoid creating silos of functionality that is only compatible with one framework.

**Reply by sidkshatriya** [reply]
Easy implementation of form validation (and error handling), dealing with complex forms without much custom code, possible support for multipage forms, some ajaxification of form validation... these are some of the kind of things I'm looking for. Its not necessary for all these features to be present.

Essentially, I'd like to have the framework (or a third party library) deal with the messy aspects of form handling. Sorry if this is a vague specification...

**Reply by davidpdrsn** [reply]
I'm sorry but the reasons I mentioned above these things are out of scope for axum, at least for now. I think you'll either have to find another library (or write one) that has what you're looking and integrate that with axum through `FromRequest` and `IntoResponse`.

**Reply by sidkshatriya** [reply]
I can appreciate if this is currently out of scope of axum.  For template handling for instance there is askama...are you aware of some generic form handling libraries out there that are good?

Thanks for your help!

**Reply by davidpdrsn** [reply]
Thanks for understanding 😊

Unfortunately not. I have only written JSON APIs in Rust where I did the validation manually. Might be something on https://www.arewewebyet.org

Please share it here if you find something that works well.

**Reply by jplatte** [reply]
The one crate im aware of in this space is [validator](https://docs.rs/validator/0.14.0/validator/).

**Reply by sidkshatriya** [reply]
@jplatte Thanks for pointing me to validator. Useful. 

Of course, there will be the issue of sending back the form back to the user with the error messages inlined which will need to be done with custom code / some other third party library...
